### PR TITLE
test.py: handle broken clusters for Python suite

### LINKED ISSUE #12360: test.py: after-test check may fail to detect a crashed instance in some cases
In scylla-enterprise branch, we have the following test.py.log:
```
------ Ending test test_clustering_order.1 ------
------ Starting test test_describe.2 ------
INFO  2022-12-19 12:55:33,754 [shard 0] migration_manager - Create new Keyspace: KSMetaData{name=cql_test_1671447333753, strategyClass=org.apache.cassandra.locator.NetworkTopologyStrategy, strategyOptions={datacenter1=1}, cfMetaData={}, durable_writes=1, userTypes=org.apache.cassandra.config.UTMetaData@0x6000006bc1a0}
INFO  2022-12-19 12:55:33,756 [shard 0] schema_tables - Creating keyspace cql_test_1671447333753
WARN  2022-12-19 12:55:33,758 [shard 0] storage_service - Failure to update pending ranges for keyspace cql_test_1671447333753 ignored
INFO  2022-12-19 12:55:33,775 [shard 0] schema_tables - Schema version changed to 481b1ada-e093-35a2-a986-5e788cb4e0e5
scylla: seastar/include/seastar/core/sharded.hh:796: Service &seastar::sharded<audit::audit>::local() [Service = audit::audit]: Assertion `local_is_initialized()' failed.
Aborting on shard 0.
Backtrace:
  0x3630f78
  0x36591d2
  /lib64/libc.so.6+0x3cb4f
  /lib64/libc.so.6+0x8ce7b
  /lib64/libc.so.6+0x3caa5
  /lib64/libc.so.6+0x267fb
  /lib64/libc.so.6+0x2671a
  /lib64/libc.so.6+0x35655
  0x24140f8
  0x2412f64
  0x1d08338
  0x1d125d7
  0x1d005d8
  0x185dbe2
  0x187237b
  0x185d7bd
  0x18597be
  0x1886471
  0x1887c2b
  0x1887ab9
  0x18877b4
  0x35fb613
  0x363eab4
  0x363fba7
  0x363f059
  0x35f34c3
  0x35f2738
  0x10356f4
  0x1064f60
  0x10349a7
  /lib64/libc.so.6+0x2750f
  /lib64/libc.so.6+0x275c8
  0x1033ce4
------ Starting test test_filtering.2 ------
------ Starting test test_frozen_collection.2 ------
------ Starting test test_json.3 ------
------ Starting test test_large_cells_rows.2 ------
------ Starting test test_logs.1 ------
------ Starting test test_lwt.1 ------
------ Starting test test_materialized_view.1 ------
------ Starting test test_minmax.3 ------
------ Starting test test_non_deterministic_functions.3 ------
------ Starting test test_null.3 ------
------ Starting test test_range_and_slice.1 ------
------ Starting test test_scan.1 ------
------ Starting test test_secondary_index.1 ------
------ Starting test test_service_levels.1 ------
------ Starting test test_shedding.1 ------
------ Starting test test_ssl.1 ------
------ Starting test test_sstable.1 ------
------ Starting test test_static.1 ------
------ Starting test test_system_tables.1 ------
------ Starting test test_tombstone_limit.2 ------
------ Starting test test_tools.2 ------
------ Starting test test_ttl.2 ------
------ Starting test test_type_duration.2 ------
------ Starting test test_type_string.3 ------
------ Starting test test_type_time.3 ------
------ Starting test test_uda.1 ------
------ Starting test test_use.2 ------
------ Starting test test_using_timeout.2 ------
------ Starting test test_utf8.2 ------
------ Starting test test_validation.2 ------
------ Starting test test_virtual_tables.3 ------
```

This shows that test.py keeps trying to use a crashed instance for a test.

How to reproduce:

It's unclear, but let's start by adding an abort() call to e.g. storage_proxy::cas, and issuing a CAS from a test. This would lead to a server crashing on any test that has an LWT in it -and not crashing in tests which don't use LWT, so not crashing right away. This server should not be returned to the pool.

### PR DESCRIPTION
**Pull request by alecco**
If the after test check fails (is_after_test_ok is False), discard the cluster and raise exception so context manager (pool) does not recycle it.

Ignore exception re-raised by the context manager.

Fixes #12360

### CODE REVIEW

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/ea98c1a6155ce5def6d9f05e5351ba9d6c08b792:

- make `Pool` context manager handle exception

File: test/pylib/pool.py [hunk 1]
```diff
@@ -97,7 +97,7 @@ async def __aenter__(self):
                 return self.obj
 
             async def __aexit__(self, exc_type, exc, obj):
-                if self.obj:
+                if exc_type is None and self.obj:
```
Comment (kbr-scylla) [inline test/pylib/pool.py:100]: The entire point of context managers is to release resources even in presence of exceptions. This makes this context manager act in a confusing way, unlike a context manager should.

We should stop using a context manager if we want to introduce complex if-logic for releasing or not releasing resources at the end.

**Review by kbr-scylla — CHANGES_REQUESTED** [review]

File: test.py [hunk 2]
```diff
@@ -842,26 +842,37 @@ async def run(self, options: argparse.Namespace) -> Test:
...
+                except Exception as e:
+                    self.server_log = cluster.read_server_log()
+                    self.server_log_filename = cluster.server_log_filename()
+                    if self.is_before_test_ok is False:
+                        print("Test {} pre-check failed: {}".format(self.name, str(e)))
+                        print("Server log of the first server:\n{}".format(self.server_log))
+                        # Don't try to continue if the cluster is broken
+                        raise
+                    if self.is_after_test_ok is False:
+                        print("Test {} pre-check failed: {}".format(self.name, str(e)))
```
Comment (nyh) [inline test.py:866]: This message should mention post-check, not pre-check

**Comment by alecco** [comment]
Changes in https://github.com/scylladb/scylladb/commit/6d460badd09971f858b18e6616fac9b3ec222ecd:

- Custom exception `Pool.PoolDiscardItem`

File: test.py [hunk 3]
```diff
@@ -842,26 +842,37 @@ async def run(self, options: argparse.Namespace) -> Test:
...
+                        print("Server log of the first server:\n{}".format(self.server_log))
+                        logger.info(f"Discarding cluster after failed test %s...", self.name)
+                        await self.suite.clusters.steal()
```
Comment (michoecho) [inline test.py:869]: The fact that you are calling `steal()` on `clusters`, and not on the context manager, breaks the RAII abstraction suggested by the context manager. The `steal()` call should be done by the context manager.

Besides the abstraction, requiring both the `steal()` and the `raise` is error-prone — it's easy to forget the `steal`. They should happen together.

File: test/pylib/pool.py [hunk 4]
```diff
@@ -97,7 +101,7 @@ async def __aenter__(self):
                 return self.obj
 
             async def __aexit__(self, exc_type, exc, obj):
-                if self.obj:
+                if exc_type is not PoolDiscardItem and self.obj:
```
Comment (michoecho) [inline test/pylib/pool.py:104]: I don't like the idea of using special exceptions for communicating with a context manager. In particular it prevents e.g. nested `try` blocks or nested uses of the context manager. If you want to communicate with the context manager, you can just add a method to the context manager and call it, i.e:
```python
async with (cm := self.suite.clusters.instance(logger)) as cluster:
     ...
     if cluster_broken:
          cm.discard()
     ...
```

File: test.py (hunk 3 above)
Comment (michoecho) [reply test.py:869]: Maybe a regular `try...except`, not RAII, could be the better abstraction here.

**Review by michoecho — CHANGES_REQUESTED** [review]

File: test/pylib/pool.py (hunk 1 above)
Comment (alecco) [reply test/pylib/pool.py]: What do you think of a  `Pool.destroy()` symmetric to `build()`.

**Review by nyh — APPROVED** [review]

**Pull request merged**